src/mscorlib/src/System/Collections/Generic/EqualityComparer.cs

```diff
@@ -36,57 +36,37 @@ private static EqualityComparer<T> CreateComparer()
             Contract.Ensures(Contract.Result<EqualityComparer<T>>() != null);
             
             object result = null;
-            RuntimeType t = (RuntimeType)typeof(T);
+            var type = (RuntimeType)typeof(T);
             
-            // Specialize type byte for performance reasons
-            if (t == typeof(byte)) {
+            // Specialize for byte so Array.IndexOf is faster.
+            if (type == typeof(byte))
```

Comment:
There is still a lot of boiled plate code that will be duplicated for every valuetype instantiation. I think you should be able to actually move all this whole method into non-generic type. The only generic part should be:

`static readonly Comparer<T> defaultComparer = (Comparer<T>)EqualityComparerHelpers.GetComparer(typeof(T), typeof(IEquatable<T>));`

(Getting rid of `IEquatable<T>` is possible but a bit harder - you would need to check whether the type implement `IEquatable<T>` without actually loading the `IEquatable<T>` type. It can be done later.)